tests/filtered_relation/tests.py

```diff
@@ -0,0 +1,346 @@
+                book_alice=FilteredRelation(
+                    'book', condition=~Q(book__title__icontains='alice'),
+                )).filter(book_alice__isnull=False).distinct(),
+            ["<Author: Jane>"])
+
+    def test_filtered_relation_with_m2m(self):
+        qs = Author.objects.annotate(
+            favourite_books_written_by_jane=FilteredRelation(
+                'favourite_books', condition=Q(favourite_books__in=[self.book2]),
+            )).filter(favourite_books_written_by_jane__isnull=False)
+        self.assertQuerysetEqual(qs, ["<Author: Alice>"])
+
+    def test_filtered_relation_with_m2m_deep(self):
+        qs = Author.objects.annotate(
+            favourite_books_written_by_jane=FilteredRelation(
+                'favourite_books', condition=Q(favourite_books__author=self.author2),
+            )).filter(favourite_books_written_by_jane__title='The book by Jane B')
+        self.assertQuerysetEqual(qs, ["<Author: Alice>"])
+
+    def test_filtered_relation_with_m2m_multijoin(self):
+        qs = Author.objects.annotate(
+            favourite_books_written_by_jane=FilteredRelation(
+                'favourite_books', condition=Q(favourite_books__author=self.author2),
+            )).filter(favourite_books_written_by_jane__editor__name='b').distinct()
+        self.assertQuerysetEqual(qs, ["<Author: Alice>"])
+
+    def test_filtered_relation_values_list(self):
+        self.assertSequenceEqual(Author.objects.annotate(
+            book_alice=FilteredRelation(
+                'book', condition=Q(book__title__iexact='poem by alice'),
+            )).values_list('book_alice__title', flat=True).order_by(F('book_alice__title').desc(nulls_last=True)),
+            ["Poem by Alice", None])
+
+    def test_filtered_relation_values(self):
+        self.assertSequenceEqual(Author.objects.annotate(
+            book_alice=FilteredRelation(
+                'book', condition=Q(book__title__iexact='poem by alice'),
+            )).filter(book_alice__isnull=False).values(),
+            [{'id': 1, 'name': 'Alice'}])
+
+    def test_filtered_relation_extra(self):
+        self.assertQuerysetEqual(Author.objects.annotate(
+            book_alice=FilteredRelation(
+                'book', condition=Q(book__title__iexact='poem by alice'),
+            )).extra(where=["1 = 1"]).order_by(
+            F('book_alice__id').desc(nulls_last=True)),
+            ["<Author: Alice>", "<Author: Jane>"])
+
+    @skipUnlessDBFeature('supports_select_union')
+    def test_filtered_relation_union(self):
+        qs1 = Author.objects.annotate(
+            book_alice=FilteredRelation(
+                'book', condition=Q(book__title__iexact='poem by alice'),
+            )).filter(book_alice__isnull=False).order_by()
+        qs2 = Author.objects.annotate(
+            book_jane=FilteredRelation(
+                'book', condition=Q(book__title__iexact='the book by jane a'),
+            )).filter(book_jane__isnull=False).order_by()
```

Comment:
the two `order_by()`s here and above are noise in the tests to work around the default ordering on the model, and they're repeated below. I think this highlights Tim's above comment about not using a default ordering and instead using `assertCountEqual`